How many late days are you using for this assignment? 

In [77]:
# I am using 0 late days. 

# Imports and pip installations (if needed)
import pandas as pd
import numpy as np
from sklearn.preprocessing import MinMaxScaler, OneHotEncoder
import matplotlib.pyplot as plt

# Part 1: Load the dataset

In [78]:
# Load the given datasets
df1 = pd.read_csv("chronic_kidney_disease_categorical.csv")  
df2 = pd.read_csv("chronic_kidney_disease_numerical.csv")

# Print the data
print("Dataset 1:")
print(df1) #categorical: 430
print("Dataset 2:")
print(df2) #numerical: 420


Dataset 1:
     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  good  yes

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: The 2 datasets have different types of information about the same patients/entries suffering from chronic kidney disease. 

In Dataset 1, the categorical set, it has features where most sections are answered with "yes/no" or "present/notPresent." It contains the sections: unique_id, albumin, sugar, red blood cells, pus cell, pus cell clumps, bacteria, hypertension, diabetes mellitus, coronary artery disease, appetite, pedal edema, anemia, and Target which says if the patient has "ckd" or "notckd"

Dataset 2 is the numerical set with sections: unique_id, age, blood pressure, blood glucose random, blood urea, serum creatinine, sodium, potassium, hemoglobin, packed cell volume, white blood cell count, and red blood cell count. These labels of the sections contain data in numbers, rather than yes/no in the categorical set.
3 examples of categorical columns: 
1. bacteria 
2. hypertension
3. diabetes mellitus

3 examples of numerical columns:

1. age
2. sodium
3. hemoglobin

Some of the columns are both categorical and numerical such as red blood cell, which is either abnormal or normal, and red blood cell count, which is a number.

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [79]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
numerical_duplicates = df2.duplicated(subset=['unique_id']).sum()
print(f"Total number of duplicates in numerical dataset: {numerical_duplicates}")

# Drop these duplicate rows
df2 = df2.drop_duplicates(subset=['unique_id'], keep='first')

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
categorical_duplicates = df1.duplicated(subset=['unique_id']).sum()
print(f"Total number of duplicate rows in categorical dataset: {categorical_duplicates}")

# Drop duplicate rows
df1 = df1.drop_duplicates(subset=['unique_id'], keep='first')

##works
# left w 400 unique patients in df1 aand 390 unique patients in df2
# 10 patients were present in df1 but had no corresponding numerical 

Total number of duplicates in numerical dataset: 30
Total number of duplicate rows in categorical dataset: 30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: Not all unique ids are present in both of the datasets. It seems that all of the unique ids/patients in the numerical dataset corresponds to a patient in the categorical dataset; however, there are 10 patients that exist in the categorical dataset, but don't exist in the numerical dataset. I know this by using print statements: 

print(f"Unique IDs in numerical dataset but missing in categorical dataset: {len(missing_in_df1)}")
print(f"Unique IDs in categorical dataset but missing in numerical dataset: {len(missing_in_df2)}")

When we merge the datasets, the rows that are missing from the numerical dataset will probably have categorical information, but their numerical data will say "NaN."


In [80]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
unique_ids_df1 = set(df1['unique_id'])
unique_ids_df2 = set(df2['unique_id'])

#check for missing unique ids in both datasets
missing_in_df1 = unique_ids_df2 - unique_ids_df1
missing_in_df2 = unique_ids_df1 - unique_ids_df2

combined_df = pd.merge(df1, df2, on='unique_id', how='outer')

#Print the combined dataset
print("\nCombined dataset:")
print(combined_df)


Combined dataset:
     unique_id   al   su     rbc        pc         pcc          ba  htn   dm  \
0       102677  3.0  0.0  normal  abnormal  notpresent  notpresent  yes   no   
1       103570  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
2       104675  2.0  1.0  normal    normal     present  notpresent   no  yes   
3       105923  1.0  0.0     NaN    normal  notpresent  notpresent  yes  yes   
4       107297  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
..         ...  ...  ...     ...       ...         ...         ...  ...  ...   
395     995177  4.0  3.0  normal  abnormal     present     present  yes  yes   
396     996076  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
397     996412  NaN  NaN     NaN       NaN  notpresent  notpresent   no   no   
398     997629  0.0  0.0  normal    normal         NaN         NaN   no   no   
399     997876  0.0  0.0     NaN    normal  notpresent  notpresent   no   no   

     cad  ...    bp 

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [81]:
# Calculate the percentage of rows that contain atleast one missing value
total_rows = combined_df.shape[0]
missing_rows = combined_df.isnull().any(axis=1).sum()
print(f"Total rows with at least one missing value: {missing_rows}")
percentage_missing = (missing_rows / total_rows) * 100
# Print %
print(f"Percentage of rows with at least 1 missing value: {percentage_missing: .2f}%")
# Drop these rows from the dataset
cleaned_combined_df = combined_df.dropna()
# Print the Dataset
print(cleaned_combined_df) #should have 153 rows because 247 entired had a missing value our of 400

Total rows with at least one missing value: 247
Percentage of rows with at least 1 missing value:  61.75%
     unique_id   al   su     rbc        pc         pcc          ba  htn   dm  \
0       102677  3.0  0.0  normal  abnormal  notpresent  notpresent  yes   no   
6       109053  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
7       114717  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
8       118191  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
9       118596  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
..         ...  ...  ...     ...       ...         ...         ...  ...  ...   
390     980291  3.0  0.0  normal  abnormal  notpresent  notpresent  yes  yes   
391     991031  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
394     994377  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
395     995177  4.0  3.0  normal  abnormal     present     present  yes  yes   
396     996076

## Part 3.4: Sort the dataset according to the Labels

In [82]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sorted_df = cleaned_combined_df.sort_values(by='Target').reset_index(drop=True)

# Print the dataset
print("\nSorted Dataset according to Target column:")
print(sorted_df)


Sorted Dataset according to Target column:
     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       102677  3.0  0.0    normal  abnormal  notpresent  notpresent  yes   
1       337834  4.0  0.0    normal  abnormal     present  notpresent  yes   
2       343710  2.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       349892  1.0  3.0  abnormal  abnormal  notpresent  notpresent  yes   
4       353025  3.0  4.0    normal    normal  notpresent  notpresent  yes   
..         ...  ...  ...       ...       ...         ...         ...  ...   
148     486397  0.0  0.0    normal    normal  notpresent  notpresent   no   
149     203466  0.0  0.0    normal    normal  notpresent  notpresent   no   
150     482105  0.0  0.0    normal    normal  notpresent  notpresent   no   
151     228849  0.0  0.0    normal    normal  notpresent  notpresent   no   
152     996076  0.0  0.0    normal    normal  notpresent  notpresent   no   

      dm  cad  ...    bp    bgr

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [83]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']

encoded_df = pd.get_dummies(sorted_df, columns=categorical_columns, drop_first=False)
encoded_df.drop(columns=['Target_notckd'], inplace=True)

# Print the dataset
print("\nDataset after encoding categorical dataset:")
print(encoded_df.columns)


Dataset after encoding categorical dataset:
Index(['unique_id', 'al', 'su', 'age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot',
       'hemo', 'pcv', 'wbcc', 'rbcc', 'rbc_abnormal', 'rbc_normal',
       'pc_abnormal', 'pc_normal', 'pcc_notpresent', 'pcc_present',
       'ba_notpresent', 'ba_present', 'htn_no', 'htn_yes', 'dm_no', 'dm_yes',
       'cad_no', 'cad_yes', 'appet_good', 'appet_poor', 'pe_no', 'pe_yes',
       'ane_no', 'ane_yes', 'Target_ckd'],
      dtype='object')


In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: If the categorical variable only had 2 values, yes/no, using a single column would be equivalent. 
If the categorical variable had more than 2 values, a single column would not work. For this example, if we have 1 column, the model may perceive the numbering of condiments as rankings, when it could just be randomly assigned. It also hinders predictability of the ratings of a sandwich because you aren't seeing how much each condiments factors into rating.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [84]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

# Print the dataset
cleaned_outliers_df = encoded_df.copy()
for column in numerical_columns:
    mean = cleaned_outliers_df[column].mean()
    std_dev = cleaned_outliers_df[column].std()

    lower_bound = mean - 3 * std_dev
    upper_bound = mean + 3 * std_dev

    cleaned_outliers_df = cleaned_outliers_df[(cleaned_outliers_df[column] >= lower_bound) & (cleaned_outliers_df[column] <= upper_bound)]

cleaned_outliers_df = cleaned_outliers_df.reset_index(drop=True) 
print("\nDataset after removing outliers:")
print(cleaned_outliers_df) #20 rows were outliers and removed


Dataset after removing outliers:
     unique_id   al   su   age     bp    bgr     bu   sc    sod  pot  ...  \
0       343710  2.0  0.0  61.0   80.0  173.0  148.0  3.9  135.0  5.2  ...   
1       397388  2.0  2.0  63.0  100.0  280.0   35.0  3.2  143.0  3.5  ...   
2       438182  2.0  0.0  73.0   80.0  253.0  142.0  4.6  138.0  5.8  ...   
3       474407  4.0  0.0  21.0   90.0  107.0   40.0  1.7  125.0  3.5  ...   
4       475200  4.0  2.0  64.0  100.0  163.0   54.0  7.2  140.0  4.6  ...   
..         ...  ...  ...   ...    ...    ...    ...  ...    ...  ...  ...   
128     486397  0.0  0.0  20.0   70.0  123.0   44.0  1.0  135.0  3.8  ...   
129     203466  0.0  0.0  22.0   60.0   97.0   18.0  1.2  138.0  4.3  ...   
130     482105  0.0  0.0  44.0   60.0   95.0   46.0  0.5  138.0  4.2  ...   
131     228849  0.0  0.0  58.0   80.0  100.0   50.0  1.2  140.0  3.5  ...   
132     996076  0.0  0.0  57.0   60.0  105.0   49.0  1.2  150.0  4.7  ...   

     dm_yes  cad_no  cad_yes  appet_good 

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [85]:
# Normalize the all Numerical Attributes in the dataset.
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

normalized_df = cleaned_outliers_df.copy()

scaler = MinMaxScaler()

normalized_df[numerical_columns] = scaler.fit_transform(normalized_df[numerical_columns])
# Print the dataset
print("\nNormalized Dataset:")
print(normalized_df)


Normalized Dataset:
     unique_id   al   su       age    bp       bgr        bu        sc  \
0       343710  2.0  0.0  0.743243  0.50  0.442060  0.901961  0.432099   
1       397388  2.0  2.0  0.770270  1.00  0.901288  0.163399  0.345679   
2       438182  2.0  0.0  0.905405  0.50  0.785408  0.862745  0.518519   
3       474407  4.0  0.0  0.202703  0.75  0.158798  0.196078  0.160494   
4       475200  4.0  2.0  0.783784  1.00  0.399142  0.287582  0.839506   
..         ...  ...  ...       ...   ...       ...       ...       ...   
128     486397  0.0  0.0  0.189189  0.25  0.227468  0.222222  0.074074   
129     203466  0.0  0.0  0.216216  0.00  0.115880  0.052288  0.098765   
130     482105  0.0  0.0  0.513514  0.00  0.107296  0.235294  0.012346   
131     228849  0.0  0.0  0.702703  0.50  0.128755  0.261438  0.098765   
132     996076  0.0  0.0  0.689189  0.00  0.150215  0.254902  0.098765   

          sod       pot  ...  dm_yes  cad_no  cad_yes  appet_good  appet_poor  \
0    0.50

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: The 'unique_id' column is not necessary for appropriate for modeling or predictions because it is merely identifying the different patients. It has no effect on if they have chronic kidney disease or not, so the model doesn't learn anything bu simply known the patients id number. 

In [86]:
#Remove that column
final_df = normalized_df.drop(columns=['unique_id'])
# Print the dataset
print("\nDataset after removing unnecessary columns:")
print(final_df)


Dataset after removing unnecessary columns:
      al   su       age    bp       bgr        bu        sc       sod  \
0    2.0  0.0  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
1    2.0  2.0  0.770270  1.00  0.901288  0.163399  0.345679  0.766667   
2    2.0  0.0  0.905405  0.50  0.785408  0.862745  0.518519  0.600000   
3    4.0  0.0  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
4    4.0  2.0  0.783784  1.00  0.399142  0.287582  0.839506  0.666667   
..   ...  ...       ...   ...       ...       ...       ...       ...   
128  0.0  0.0  0.189189  0.25  0.227468  0.222222  0.074074  0.500000   
129  0.0  0.0  0.216216  0.00  0.115880  0.052288  0.098765  0.600000   
130  0.0  0.0  0.513514  0.00  0.107296  0.235294  0.012346  0.600000   
131  0.0  0.0  0.702703  0.50  0.128755  0.261438  0.098765  0.666667   
132  0.0  0.0  0.689189  0.00  0.150215  0.254902  0.098765  1.000000   

          pot      hemo  ...  dm_yes  cad_no  cad_yes  appet_good  appet_poor 

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [87]:
# Export the dataframe to a new csv file
final_df.to_csv("cleaned_kidney_disease.csv", index=False)
# Export the dataframe to a new json file
final_df.to_json("cleaned_kidney_disease.json", orient="records", indent=4)
print("Data successfully exported to 'cleaned_kidney_disease.csv' and 'cleaned_kidney_disease.json'")

Data successfully exported to 'cleaned_kidney_disease.csv' and 'cleaned_kidney_disease.json'


# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [88]:
print("\nTop 15 rows of the cleaned dataset:")
print(final_df.head(15))


Top 15 rows of the cleaned dataset:
     al   su       age    bp       bgr        bu        sc       sod  \
0   2.0  0.0  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
1   2.0  2.0  0.770270  1.00  0.901288  0.163399  0.345679  0.766667   
2   2.0  0.0  0.905405  0.50  0.785408  0.862745  0.518519  0.600000   
3   4.0  0.0  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
4   4.0  2.0  0.783784  1.00  0.399142  0.287582  0.839506  0.666667   
5   3.0  1.0  0.729730  0.00  0.935622  0.169935  0.160494  0.333333   
6   2.0  0.0  0.675676  0.75  0.253219  0.633987  0.777778  0.366667   
7   4.0  3.0  0.851351  0.25  0.618026  0.562092  0.728395  0.000000   
8   1.0  0.0  0.540541  0.00  0.399142  0.535948  0.358025  0.700000   
9   3.0  0.0  0.756757  0.25  0.223176  0.209150  0.160494  0.533333   
10  3.0  1.0  0.662162  0.50  0.618026  0.411765  0.432099  0.566667   
11  4.0  1.0  0.783784  0.00  0.725322  0.313725  0.481481  0.566667   
12  4.0  0.0  0.878378  0.0

Paste here: 

| al  | su  | age     | bp   | bgr     | bu      | sc      | sod    | pot     | hemo    | pcv     | wbcc    | rbcc    | pc_normal | pcc_present | ba_present | htn_yes | dm_yes | cad_yes | appet_poor | pe_yes | ane_yes | Target_notckd |
|---- |---- |-------- |----  |-------- |-------- |-------- |--------|-------- |-------- |-------- |-------- |-------- |-----------|-------------|------------|---------|--------|-------- |------------|--------|-------- |-------------- |
| 2.0 | 0.0 | 0.743243 | 0.50 | 0.442060 | 0.901961 | 0.432099 | 0.500000 | 0.793103 | 0.000000 | 0.425743 | 0.321678 | 0.435897 | False     | False       | False      | True    | True   | True    | True       | True   | True   | False         |
| 2.0 | 2.0 | 0.770270 | 1.00 | 0.901288 | 0.163399 | 0.345679 | 0.766667 | 0.206897 | 0.524752 | 0.376238 | 0.335664 | 0.487179 | True      | False       | True       | True    | False  | True    | False      | False  | False  | False         |
| 2.0 | 0.0 | 0.905405 | 0.50 | 0.785408 | 0.862745 | 0.518519 | 0.600000 | 1.000000 | 0.277228 | 0.257426 | 0.281118 | 0.384615 | False     | False       | False      | True    | True   | True    | False      | False  | False  | False         |
| 4.0 | 0.0 | 0.202703 | 0.75 | 0.158798 | 0.196078 | 0.160494 | 0.166667 | 0.206897 | 0.059406 | 0.415842 | 0.216783 | 0.358974 | False     | True        | True       | False   | False  | False   | False      | False  | True   | False         |
| 4.0 | 2.0 | 0.783784 | 1.00 | 0.399142 | 0.287582 | 0.839506 | 0.666667 | 0.586207 | 0.019802 | 0.326733 | 0.279503 | 0.307692 | False     | False       | True       | True    | True   | False   | False      | True   | False  | False         |
| 3.0 | 1.0 | 0.729730 | 0.00 | 0.935622 | 0.169935 | 0.160494 | 0.333333 | 0.034483 | 0.019802 | 0.287129 | 0.198129 | 0.205128 | False     | True        | False      | True    | False  | False   | True       | False  | True   | False         |
| 2.0 | 0.0 | 0.675676 | 0.75 | 0.253219 | 0.633987 | 0.777778 | 0.366667 | 0.655172 | 0.138614 | 0.495050 | 0.306230 | 0.410256 | False     | False       | False      | True    | False  | False   | False      | False  | False  | False         |
| 4.0 | 3.0 | 0.851351 | 0.25 | 0.618026 | 0.562092 | 0.728395 | 0.000000 | 0.344828 | 0.168317 | 0.267327 | 0.259915 | 0.282051 | False     | True        | True       | True    | True   | True    | False      | True   | True   | False         |
| 1.0 | 0.0 | 0.540541 | 0.00 | 0.399142 | 0.535948 | 0.358025 | 0.700000 | 0.379310 | 0.207921 | 0.376238 | 0.323684 | 0.410256 | True      | False       | False      | True    | True   | False   | False      | False  | False  | False         |
| 3.0 | 0.0 | 0.756757 | 0.25 | 0.223176 | 0.209150 | 0.160494 | 0.533333 | 0.620690 | 0.485149 | 0.405941 | 0.293384 | 0.307692 | False     | False       | False      | True    | True   | False   | False      | False  | False  | False         |
| 3.0 | 1.0 | 0.662162 | 0.50 | 0.618026 | 0.411765 | 0.432099 | 0.566667 | 0.689655 | 0.316832 | 0.425743 | 0.320164 | 0.410256 | False     | True        | True       | True    | True   | False   | False      | True   | False  | False         |
| 4.0 | 1.0 | 0.783784 | 0.00 | 0.725322 | 0.313725 | 0.481481 | 0.566667 | 0.862069 | 0.178218 | 0.336634 | 0.288748 | 0.333333 | False     | False       | True       | True    | True   | False   | True       | True   | False  | False         |
| 4.0 | 0.0 | 0.878378 | 0.00 | 0.206009 | 0.751634 | 0.604938 | 0.533333 | 0.689655 | 0.366337 | 0.356436 | 0.273083 | 0.358974 | True      | False       | False      | True    | True   | False   | True       | True   | False  | False         |
| 3.0 | 0.0 | 0.878378 | 0.25 | 0.639485 | 0.470588 | 0.395062 | 0.433333 | 0.517241 | 0.267327 | 0.297030 | 0.278502 | 0.307692 | False     | True        | True       | True    | True   | True    | False      | False  | False  | False         |
| 1.0 | 0.0 | 0.716216 | 0.50 | 1.000000 | 0.163399 | 0.111111 | 0.066667 | 0.206897 | 0.267327 | 0.366337 | 0.285452 | 0.205128 | True      | False       | False      | False   | True   | False   | True       | False  | False  | False         |



### Paste the markdown table here

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

Answer: Keeping the input size small could help with conversion hallucinations so there isn't too much data to process. Additionally, it is important to be specific with what I am prompting the LLM and ensure that the info input is correct and what we want to use. 

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [89]:
# Code to rename all the columns in the dataset
column_mapping = {
    'age': 'Age',
    'bp': 'Blood Pressure',
    'sg': 'Specific Gravity',
    'al': 'Albumin',
    'su': 'Sugar',
    'rbc': 'Red Blood Cells',
    'pc': 'Pus Cell',
    'pcc': 'Pus Cell Clumps',
    'ba': 'Bacteria',
    'bgr': 'Blood Glucose Random',
    'bu': 'Blood Urea',
    'sc': 'Serum Creatinine',
    'sod': 'Sodium',
    'pot': 'Potassium',
    'hemo': 'Hemoglobin',
    'pcv': 'Packed Cell Volume',
    'wbcc': 'White Blood Cell Count',
    'rbcc': 'Red Blood Cell Count',
    'htn': 'Hypertension',
    'dm': 'Diabetes Mellitus',
    'cad': 'Coronary Artery Disease',
    'appet': 'Appetite',
    'pe': 'Pedal Edema',
    'ane': 'Anemia',
    'class': 'Chronic Kidney Disease'
    
}

# Use the rename method to rename the columns in the DataFrame
final_df.rename(columns=column_mapping, inplace=True)

# Print the updated DataFrame to confirm the changes
print(final_df)

     Albumin  Sugar       Age  Blood Pressure  Blood Glucose Random  \
0        2.0    0.0  0.743243            0.50              0.442060   
1        2.0    2.0  0.770270            1.00              0.901288   
2        2.0    0.0  0.905405            0.50              0.785408   
3        4.0    0.0  0.202703            0.75              0.158798   
4        4.0    2.0  0.783784            1.00              0.399142   
..       ...    ...       ...             ...                   ...   
128      0.0    0.0  0.189189            0.25              0.227468   
129      0.0    0.0  0.216216            0.00              0.115880   
130      0.0    0.0  0.513514            0.00              0.107296   
131      0.0    0.0  0.702703            0.50              0.128755   
132      0.0    0.0  0.689189            0.00              0.150215   

     Blood Urea  Serum Creatinine    Sodium  Potassium  Hemoglobin  ...  \
0      0.901961          0.432099  0.500000   0.793103    0.000000  ... 

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [90]:
# Converted SQL to Pandas code
result = final_df.groupby('Target_ckd').size().reset_index(name='count')
print(result)

   Target_ckd  count
0       False    111
1        True     22


*Describe what the above code does here*

The code groups the dataset by the values in the Target_ckd column and counts them using .size(). So the output shows how many people have ckd: 22, and not ckd: 111. 